### PR TITLE
feat: always display both underlying and rewards balances in opportunity row

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Direct follow-up of https://github.com/shapeshift/web/pull/4692:

> We assume an opportunity row can have rewards nested asset rows, or underlying asset rows, not the two together. This was perfectly valid previously, but now that we abuse the staking DefiType (and will most likely use it as our only type going forward, removing DefiType altogether) for the read-only opportunities, this makes us unable to see the underlying assets.
This will need to be tackled in a separate PR - once that's done, underlying holdings should work out-of-the-box.

This PR now makes opportunity rows always display underlyings and rewards, instead of assuming one or the other.

The below ixed to our best effort with the data we get, though some assets are rugged because of wrong liquidity/supply upstream. Will investigate how to rely on `token.balance` as a first option in a follow-up (it's currently second in the op-chain [here](https://github.com/shapeshift/web/blob/7978d9fb791055e7c171606dedde818e72a21af1/src/state/apis/zapper/zapperApi.ts#L572-L573), which *would* work for the follow-up assets, though may break others)

~~Note, underlyings may be wrong and not always make up for the total currently (i.e sKLIMA, sOHM in willy's wally), while this isn't directly related to this view-layer change per se, it will need to be fixed as part of this PR before opening it so it actually makes sense when reviewing it with develop flags.~~

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Given our underlying assets logic, the underlying we display for current opportunities may be wrong - test accordingly

## Screenshots (if applicable)

- Current opportunities

image
image
image
image
image
image

- Read-only opportunities

image
image
image
image
image
image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4680** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4680" target="_blank">Graphite</a> 
    * **PR #4679** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4679" target="_blank">Graphite</a> 
      * **PR #4681** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4681" target="_blank">Graphite</a> 
        * **PR #4682** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4682" target="_blank">Graphite</a> 
          * **PR #4690** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4690" target="_blank">Graphite</a> 
            * **PR #4691** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4691" target="_blank">Graphite</a> 
              * **PR #4692** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4692" target="_blank">Graphite</a> 
                * **PR #4693** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4693" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4693?utm_source=stack-comment).

File: src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx [hunk 1]
```diff
@@ -130,29 +124,43 @@ export const OpportunityRow: React.FC<
...
+          if (bnOrZero(underlyingAssetBalance.cryptoBalancePrecision).isZero()) return null
           return (
             <NestedAsset
               key={assetId}
               isClaimableRewards={isClaimableRewards}
               isExternal={opportunity.isReadOnly}
               assetId={assetId}
-              balances={underlyingAssetBalances[assetId]}
+              balances={underlyingAssetBalance}
               onClick={() => handleClick(DefiAction.Claim)}
```
Comment (gomesalexandre) [inline src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx:150]: Obviously wrong for underlying assets

File: src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx [hunk 2]
```diff
@@ -126,34 +110,51 @@ export const OpportunityRow: React.FC<
     ))
   }, [apy, cryptoAmountBaseUnit, group, type])
 
-  const renderRewardAssets = useMemo(() => {
```
Comment (gomesalexandre) [inline src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx:129]: This wasn't rendering reward assets only, but underlying or rewards

File: src/state/apis/zapper/zapperApi.ts [hunk 3]
```diff
@@ -566,20 +567,25 @@ export const zapper = createApi({
...
-                      const reserveBaseUnit = toBaseUnit(reserve, asset.decimals ?? 18)
-                      const totalSupplyBaseUnit =
-                        typeof asset.supply === 'number'
-                          ? toBaseUnit(asset.supply, asset.decimals ?? 18)
-                          : undefined
-                      const tokenPoolRatio = totalSupplyBaseUnit
-                        ? bn(reserveBaseUnit).div(totalSupplyBaseUnit).toString()
+                  const underlyingAssetRatiosBaseUnit = (
+                    asset.dataProps?.reserves ??
+                    topLevelAsset.tokens?.map(
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/apis/zapper/zapperApi.ts:572]: Disregard me, dis happy

**Review by kaladinlight — DISMISSED** [review]

File: src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx [hunk 4]
```diff
@@ -126,34 +112,51 @@ export const OpportunityRow: React.FC<
...
+            // Don't display the same asset as an underlying, that's an implementation detail, but shouldn't be user-facing
+            if (underlyingAssetId === assetId) return null
+            if (bnOrZero(underlyingAssetBalance.cryptoBalancePrecision).isZero()) return null
+            return (
+              <NestedAsset
+                key={underlyingAssetId}
+                isClaimableRewards={isClaimableRewards}
+                assetId={underlyingAssetId}
+                balances={underlyingAssetBalance}
+                onClick={() => history.push(`/assets/${underlyingAssetId}`)}
```
Comment (gomesalexandre) [inline src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx:143]: Just for the sake of making `onClick` happy (it is required) and to enable the click if we so wish in the future, at the moment, `<NestedAsset />` only makes rewards clickable, not underlying assets

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**